packetbeat/protos/dns/dns.go

```diff
@@ -501,6 +501,23 @@ func addDNSToMapStr(m common.MapStr, dns *mkdns.Msg, authority bool, additional
 			// etld_plus_one should be removed for 8.0.0.
 			qMapStr["etld_plus_one"] = eTLDPlusOne
 			qMapStr["registered_domain"] = eTLDPlusOne
+
+			if eTLDPlusOne != "" && strings.Contains(eTLDPlusOne, ".") {
+				topLevelDomain := strings.Join(strings.Split(eTLDPlusOne, ".")[1:], ".")
+				qMapStr["top_level_domain"] = topLevelDomain
+
+				subdomain := strings.Trim(strings.TrimSuffix(q.Name, eTLDPlusOne), ".")
+				if subdomain != "" {
+					qMapStr["subdomain"] = subdomain
+				}
+			}
+		} else if strings.Contains(err.Error(), "cannot derive eTLD+1 for domain") && strings.Count(q.Name, ".") == 1 {
```

Comment:
Matching error strings is really error prone because the error can change upstream without notice. Is there a way you can do this without matching the error string?